File: backend/src/schemas/dtos/availability_rule_dto.py
from typing import Optional
from datetime import datetime, time
from dataclasses import dataclass
from uuid import UUID


@dataclass
class AvailabilityRuleDTO:
  """  
  DTO for communicating an availability rule.  
  """
  id: int
  uuid: UUID
  name: str
  user_id: str
  iana_timezone: str
  prevents_booking: bool = None
  weekdays: Optional[list[int]] = None
  start_time: Optional[time] = None
  end_time: Optional[time] = None
  start_datetime: Optional[datetime] = None
  end_datetime: Optional[datetime] = None
  created_at: Optional[datetime] = None
  updated_at: Optional[datetime] = None

File: frontend/src/types/AvailabilityRuleDTO.ts
import type { TZDate } from "@date-fns/tz";

// basic dto with primative types
export type AvailabilityRuleDTO = {
  uuid: string;
  name: string;
  iana_timezone: string;
  prevents_booking?: boolean;
  weekdays?: number[];
  start_datetime?: string;
  end_datetime?: string;
  start_time: string | null;
  end_time: string | null;
};

// dto with string dates processed into Date objects
export type ProcessedAvailabilityRuleDTO = Omit<AvailabilityRuleDTO, 'start_datetime' | 'end_datetime'> & {
  start_datetime?: TZDate;
  end_datetime?: TZDate;
};
